pkg/packageinstall/packageinstall.go

```diff
@@ -201,6 +205,74 @@ func (pi *PackageInstallCR) reconcileAppWithPackage(existingApp *kcv1alpha1.App,
 	return reconcile.Result{}, nil
 }
 
+func (pi *PackageInstallCR) clusterVersionConstraintsSatisfied(pkg *datapkgingv1alpha1.Package) bool {
+	if pkg.Spec.KubernetesVersionSelection == nil || pkg.Spec.KubernetesVersionSelection.Constraints == "" {
+		return true
+	}
+	const kubernetesVersionOverrideAnnotation = "packaging.carvel.dev/ignore-kubernetes-version-selection"
+
+	_, found := pi.model.Annotations[kubernetesVersionOverrideAnnotation]
+	if found {
+		pi.log.Info("Found kubernetes version override annotation; not applying version constraints")
+		return true
+	}
+
+	vi, err := deploy.GetClusterVersion(pi.coreClient, pi.model.Spec.ServiceAccountName, pi.model.Spec.Cluster, &pi.model.ObjectMeta, pi.log)
+	if err != nil {
+		pi.log.Error(err, "Unable to retrieve cluster kubernetes version")
+		return false
+	}
+	v, err := dropPreAndBuild(vi.GitVersion)
+	if err != nil {
+		pi.log.Error(err, "Unable to parse cluster kubernetes version", "version string", vi.GitVersion)
+		return false
+	}
+	constraints, err := semver.ParseRange(pkg.Spec.KubernetesVersionSelection.Constraints)
```

Comment:
lets add a validation for constraints on a package: https://github.com/vmware-tanzu/carvel-kapp-controller/blob/e5667d3102f34e05df58e42670689b3d6ab5ee29/pkg/apiserver/apis/datapackaging/validation/validations.go#L37. with that addition, it's fine that we would ignore the error here (and log).